docs/proposals/scheduling/20220609-resource-reservation.md

```diff
@@ -0,0 +1,270 @@
+
+#### Story 3
+
+As an application administrator, I want to make the **horizontal scaling** of my app more deterministic by submitting reservations before a scale-up. Besides, I can also reserve resources after a scale-down for future demands. It is useful especially when we want a guaranteed scale-up of applications for the coming business peak.
+
+#### Story 4
+
+As a cluster administrator, I want to **pre-allocate** node resources for future usage no matter whether they are available now or not. I want to allocate the future free resources but do not disrupt the running of scheduled pods. Reservation can be made to pre-allocate resources since it makes no physical cost to the node. It may be in a `Waiting` state. When there is enough space for the reservation, it will become `Available` for the owner pods' scheduling.
+
+### API
+
+In this section, a Custom Resource Definition (CRD) named `Reservation` is proposed to allow the scheduler to reserve node resources for specific pods.
+
+![image](/docs/images/resource-reservation.svg)
+
+```go
+type Reservation struct {
+	metav1.TypeMeta `json:",inline"`
+	// A Reservation object is non-namespaced.
+	// It can reserve resources for pods of any namespace. Any affinity/anti-affinity of reservation scheduling can be
+	// specified in the pod template.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              ReservationSpec   `json:"spec,omitempty"`
+	Status            ReservationStatus `json:"status,omitempty"`
+}
+
+type ReservationSpec struct {
+	// Template defines the scheduling requirements (resources, affinities, images, ...) processed by the scheduler just
+	// like a normal pod.
+	// If the `template.spec.nodeName` is specified, the scheduler will not choose another node but reserve resources on
+	// the specified node.
+	Template *corev1.PodTemplateSpec `json:"template,omitempty"`
+	// Specify the owners who can allocate the reserved resources.
+	// Multiple owner selectors and ANDed.
+	Owners []ReservationOwner `json:"owners,omitempty"`
+	// By default, the resources requirements of reservation (specified in `template.spec`) is filtered by whether the
+	// node has sufficient free resources (i.e. ReservationRequest <  NodeFree).
+	// When `preAllocation` is set, the scheduler will skip this validation and allow overcommitment. The scheduled
+	// reservation would be waiting to be available until free resources are sufficient.
+	PreAllocation bool `json:"preAllocation,omitempty"`
+	// Time-to-Live period for the reservation.
+	// `ttl` and `expirationTimestamp` are mutually exclusive. If both `ttl` and `expirationTimestamp` are not
+	// specified, a very long TTL will be picked by default.
+	TTL *metav1.Duration `json:"ttl,omitempty"`
```

Comment:
suggest that we follow the existing definition in community: https://github.com/kubernetes/kubernetes/blob/c1e69551be1a72f0f8db6778f20658199d3a686d/cmd/kubeadm/app/apis/bootstraptoken/v1/types.go#L33-L40

Reply:
OK.